builder/builder.go

```diff
@@ -1,61 +1,136 @@
-// package builder can be used in testing to create a temporary gopath, src, 
+// package builder can be used in testing to create a temporary go module or gopath, src,
 // namespace and package directory, and populate it with source files.
 package builder
 
 import (
+	"fmt"
 	"io/ioutil"
 	"os"
+	"path"
 	"path/filepath"
 	"strings"
 
-	"path"
-
 	"github.com/dave/patsy/vos"
 	"github.com/pkg/errors"
 )
 
-// New creates a new gopath in the system temporary location, creates the src
-// dir and the namespace dir. The gopath is appended to the beginning of the 
-// existing gopath, so existing imports will still work. Remember to defer the 
-// Cleanup() method to delete the temporary files.
-func New(env vos.Env, namespace string) (*Builder, error) {
+// New creates a new temporary location, either for a go module or for a gopath root.
+// See NewGoModule or NewGoRoot for the details.
+// Remember to defer the Cleanup() method to delete the temporary files.
+func New(env vos.Env, namespace string, gomod bool) (*Builder, error) {
+	if gomod {
+		return NewGoModule(env, namespace)
+	} else {
+		return NewGoRoot(env, namespace)
+	}
+}
 
+// NewGoRoot creates a new gopath in the system temporary location, creates the src
+// dir and the namespace dir. The gopath is appended to the beginning of the
+// existing gopath, so existing imports will still work.
+// Remember to defer the Cleanup() method to delete the temporary files.
+func NewGoRoot(env vos.Env, namespace string) (*Builder, error) {
 	gopath, err := ioutil.TempDir("", "go")
 	if err != nil {
 		return nil, errors.Wrap(err, "Error creating temporary gopath root dir")
 	}
 
+	// gopath needs to match what `go list` will be returning, so eval symlinks and clean
+	gopath, err = filepath.EvalSymlinks(filepath.Clean(gopath))
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	root := filepath.Join(gopath, "src", namespace)
+
 	b := &Builder{
 		env:       env,
-		root:      gopath,
+		gopath:    gopath,
+		root:      root,
 		namespace: namespace,
+		gomod:     false,
 	}
 
 	if err := os.Mkdir(filepath.Join(gopath, "src"), os.FileMode(0777)); err != nil {
 		b.Cleanup()
 		return nil, errors.Wrap(err, "Error creating temporary gopath src dir")
 	}
-	b.env.Setenv("GOPATH", gopath+string(filepath.ListSeparator)+b.env.Getenv("GOPATH"))
-
-	if err := os.MkdirAll(filepath.Join(gopath, "src", namespace), os.FileMode(0777)); err != nil {
+	if err := os.MkdirAll(root, os.FileMode(0777)); err != nil {
 		b.Cleanup()
 		return nil, errors.Wrap(err, "Error creating temporary namespace dir")
 	}
 
+	err = b.env.Setenv("GOPATH", gopath+string(filepath.ListSeparator)+b.env.Getenv("GOPATH"))
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	// change dir to root to ensure consistent behaviour
+	_ = os.Chdir(gopath)
```

Comment:
I'm doing `os.Chdir` here because some std lib packages (like `go/build`) use `os.Getwd()` internally in gomod mode.